File: CrackTheCodingInterview/haskell/src/Chapter1/OnePointThree.hs
{-# LANGUAGE OverloadedStrings #-}

module Chapter1.OnePointThree
  ( stringToMap
  , isPermutation
  )
where

-- 1.3 Given 2 string, check if one is a permutation of the other
-- normal larmon

import qualified Data.Text                     as T
import qualified Data.Map                      as M
import           Data.Maybe
import           Data.Char


stringHash :: String -> Int
stringHash  =
  foldl (\hashValue character -> ord character + hashValue) 0 


isPermutation :: String -> String -> Bool
isPermutation firstString secondString = stringHash firstString == stringHash secondString

File: CrackTheCodingInterview/javascript/src/chapter-one/one-point-three.js
// Given 2 string check if one is a permutation of the other
// normal larmon

export const stringToValue = (string) => {
  const stringCharacters = string.split("");
  return stringCharacters.reduce((stringValue, character) => {
    return stringValue + character.charCodeAt(0);
  }, 0);
};

export const checkPermutation = (firstString, secondString) => {
  if (firstString.length !== secondString.length) {
    return false;
  } else {
    const firstStringValue = stringToValue(firstString);
    const secondStringValue = stringToValue(secondString);
    return firstStringValue === secondStringValue;
  }
};

